### PR TITLE
Barbarian Path of the Zealot

### PR DESCRIPTION
**Pull request by nickelpro**
This code doesn't work, I'm opening this draft PR hoping someone can help me troubleshoot.

Barbarians have a lot of modifications that can be applied to Rage, this tries to solve that generically and implement the Path of the Zealot feats Divine Fury and Fanatical Focus.

How this system works:
* Rage has a simple item macro that triggers each turn (also on application and removal), searches for any Barbarian subclass, and if that subclass has an item macro attached runs that item macro. The macro arguments are forwarded to the subclass macro unaltered.

* The subclass item macro knows about the feats that the subclass has available. It searches for those feats and runs their Midi-QOL workflows if found (and when appropriate).

* All of the work of a given feat is implemented on the feat item, so Fanatical Focus's AEs are implemented on the Fanatical Focus item, Divine Fury's are implemented on the Divine Fury item, etc. Running their respective Midi-QOL workflows applies them.

I have this whole system working in Foundry, the problems I'm having come with translating to the 5eJSON format (with which I'm not really familiar).

Right now:
* The only thing that's working is the Rage ItemMacro, which gets applied correctly
* Rage effects are ***not*** added to the item when the Rage classFeature is imported
* Fanatical Focus/Divine Fury effects are ***not*** added to the items when their respective subclassFeatures are imported
* The Path of the Zealot subclass ItemMacro is ***not*** applied when the subclass is imported

So obviously I've screwed some things up, any help is appreciated.

### CODE REVIEW

File: module/data/subclass/__core.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "subclass": [
+    {
+      "name": "Path of the Zealot",
+      "source": "XGE",
+      "className": "Barbarian",
+      "classSource": "PHB",
+      "subclassShortName": "Path of the Zealot",
```
Comment (TheGiddyLimit) [inline module/data/subclass/__core.json:8]:
```suggestion
      "subclassShortName": "Zealot",
```

File: module/data/subclassFeature/__core.json [hunk 2]
```diff
@@ -46,6 +46,67 @@
 					}
 				}
 			]
+		},
+		{
+			"name": "Divine Fury",
+			"source": "XGE",
+			"className": "Barbarian",
+			"classSource": "PHB",
+			"subclassShortName": "Path of the Zealot",
```
Comment (TheGiddyLimit) [inline module/data/subclassFeature/__core.json:55]:
```suggestion
			"subclassShortName": "Zealot",
```

File: module/data/subclassFeature/__core.json [hunk 3]
```diff
@@ -46,6 +46,67 @@
...
+					}
+				}
+			]
+		},
+		{
+			"name": "Fanatical Focus",
+			"source": "XGE",
+			"className": "Barbarian",
+			"classSource": "PHB",
+			"subclassShortName": "Path of the Zealot",
```
Comment (TheGiddyLimit) [inline module/data/subclassFeature/__core.json:86]:
```suggestion
			"subclassShortName": "Zealot",
```

**Review by TheGiddyLimit — CHANGES_REQUESTED** [review]
Initial eyeballing; will check out/test Soon:tm:

**Comment by TheGiddyLimit** [comment]
Pushed some fixes, but it likely still doesn't work as you'd hope/expect.

The effects _do_ get populated, but they're populated on the _actor_, not on the items themselves. Ironically, if you import the features to a compendium/the directory, I believe the effects _will_ appear on the items themselves (although I didn't test this). This is a Plutonium issue, which we can't fix here (and one I'll need to remember to fix sometime, since I can't exactly remember why it works this way!).

Also, as a side-note, running `npm t` _should_ highlight issues like the malformed subclass, but the schema isn't good enough yet (see #47 )

**Comment by TheGiddyLimit** [comment]
@nickelpro last night's Plutonium update _should_ have fixed the wonky effect adding; mind checking this out with my changes and giving it a test?

**Comment by nickelpro** [comment]
This now works, but the Divine Fury damage type is hardcoded. The normal import data uses a `chooseData` to get the user's chosen damage type (necrotic or radiant).

Ideally we would have a mechanism to override, hijack, or otherwise access that data choice from addon-automation. The normal import implements Divine Fury as a special action, which is reasonable for a non-automation build, so this isn't really something that can be moved into plutonium proper.

**Comment by TheGiddyLimit** [comment]
Hm. I could potentially make a `chooseFlags` which we could use here to allow the user to choose their damage type as before; the result would stored as a flag we could later reference:

```
"chooseFlags": [
	{
		"name": "Necrotic Damage",
		"flags": {
			"plutonium-addon-automation": {"chosenDamageType": "necrotic"}
		}
	},
	{
		"name": "Radiant Damage",
		"flags": {
			"plutonium-addon-automation": {"chosenDamageType": "radiant"}
		}
	}
]
```


```
"value": "+ (1d6 + floor(@classes.barbarian.levels / 2))[radiant]"
// becomes
"value": "+ (1d6 + floor(@classes.barbarian.levels / 2))[@item.divine-fury.flags.plutonium-addon-automation.chosenDamageType]"
```

This would require some changes in Plutonium to function:
- (Support for `chooseFlags`)
- Patching of `getRollData` to add `@item. ...`
- Ensuring `chooseData` can be sensibly overwritten (removed, here) from this module

Thoughts?

**Comment by revilowaldow** [comment]
Is it possible to further simplify this by enabling the choice at this lower level to be mapped to and then inherited from the choice made at the core level? 
Alternatively will this be able to suppress the choice in the core data so you're not asked to choose twice with your first (pointless) choice then being deleted.

**Comment by TheGiddyLimit** [comment]
> Is it possible to further simplify this by enabling the choice at this lower level to be mapped to and then inherited from the choice made at the core level?

Not easily, because the addon data here overwrites the core choice, and I _think_ we want to keep it that way. `Divine Fury` is being redefined here as a self-targeting no-damage ability, much like a (properly implemented) smite, which feels correct

> Alternatively will this be able to suppress the choice in the core data so you're not asked to choose twice with your first (pointless) choice then being deleted.

That's what my "Ensuring `chooseData` can be sensibly overwritten (removed, in this instance) from this module", above, is about. Right enough, we don't want two popups. Both `chooseData` and this hypothetical `chooseFlags` will use the same UI/etc., so there should be no visible difference to the user

**Comment by nickelpro** [comment]
> There is no problem in computer science which cannot be solved by one more level of indirection, except perhaps the problem of too many levels of indirection

My ideal solution is that we could override the original `chooseData` with a new one that chooses between the two effects, one that does necrotic and one that does radiant.

The flags solution, while flexible, sets off my complexity alarms. I think `chooseFlags` is something we might want eventually no matter what, but that damage formula is freaking me out for this particular application. Caveat: If `chooseFlags` is much easier to implement in plut, I care about a working solution more than one that suits my aesthetics.

**Comment by TheGiddyLimit** [comment]
Love a bit of indirection, me

`chooseData` is in need of a rename, now that I've dropped support for v9--it's really `chooseSystem`, now. Overwriting the original would work just fine, but it's ultimately limited to setting `system` fields on the item. Since you're overwriting the `damage.parts` to, presumably, avoid MidiQOL hitting the PC in the face with their own Divine Fury, we need somewhere else to store the data. Stuffing it into a rando `system` field feels more footgun to me than storing it in a unique flag, but if you've a better idea, I'm all ears

**Comment by nickelpro** [comment]
>  avoid MidiQOL hitting the PC in the face with their own Divine Fury

It's only funny the first couple dozen times

> it's ultimately limited to setting `system` fields on the item

Ah, limited to system fields. Nope, flags is good then, long damage formulas be damned.

**Comment by TheGiddyLimit** [comment]
I'll try get those Plutonium changes wired up Soon:tm:, and once again keep this PR open as a good test case 😏

**Comment by nickelpro** [comment]
:eyes:

**Comment by TheGiddyLimit** [comment]
Latest Plutonium should now handle the `chooseFlags`; give it a spin, if you would, @nickelpro

**Review by TheGiddyLimit — APPROVED** [review]

**Pull request merged**